edge/pkg/metamanager/metaserver/certificate/manager.go

```
@@ -0,0 +1,145 @@
+	if kubeClient != nil {
+		clientsetFn = func(current *tls.Certificate) (clientset.Interface, error) {
+			return kubeClient, nil
+		}
+	}
+
+	certificateStore, err := certificate.NewFileStore(pairNamePrefix, certDirectory, certDirectory, "", "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize server certificate store: %v", err)
+	}
+
+	getTemplate := func() *x509.CertificateRequest {
+		if len(ips) == 0 {
+			return nil
+		}
+		return &x509.CertificateRequest{
+			Subject: pkix.Name{
+				CommonName:   fmt.Sprintf("system:node:%s", nodeName),
+				Organization: []string{"system:nodes"},
+			},
+			IPAddresses: ips,
+		}
+	}
+
+	m, err := certificate.NewManager(&certificate.Config{
+		Name:        fmt.Sprintf("metaserver-csr-%s", nodeName),
+		ClientsetFn: clientsetFn,
+		GetTemplate: getTemplate,
+		SignerName:  certificates.KubeletServingSignerName,
+		Usages: []certificates.KeyUsage{
+			certificates.UsageDigitalSignature,
+			certificates.UsageKeyEncipherment,
+			certificates.UsageServerAuth,
+		},
+		CertificateStore: certificateStore,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize server certificate manager: %v", err)
+	}
+
+	return &ServerCertificateManager{m}, nil
+}
+
+func (scm *ServerCertificateManager) ready() bool {
+	if currentCert := scm.Current(); currentCert != nil {
+		return true
+	}
+	return false
+}
+
+func (scm *ServerCertificateManager) WaitForCertReady() error {
+	return wait.PollUntilContextTimeout(context.Background(), 5*time.Second, 4*time.Minute, true, func(ctx context.Context) (bool, error) {
+		if scm.ready() {
+			return true, nil
+		}
+		return false, nil
+	})
+}
+
+func (scm *ServerCertificateManager) WaitForCAReady() error {
+	return wait.PollUntilContextTimeout(context.Background(), 5*time.Second, 4*time.Minute, true, func(ctx context.Context) (bool, error) {
+		if !cloudconnection.IsConnected() {
+			return false, nil
+		}
+		err := scm.getKubeAPIServerCA()
+		if err == nil {
```

You can print the error message here, otherwise the error will be lost.

---

> You can print the error message here, otherwise the error will be lost.

Done.